gio-macros/src/lib.rs

```diff
@@ -0,0 +1,201 @@
+// Take a look at the license at the top of the repository in the LICENSE file.
+
+extern crate proc_macro;
+
+mod actions;
+
+use darling::FromMeta;
+use proc_macro::TokenStream;
+use syn::{parse_macro_input, AttributeArgs, ItemImpl};
+
+/// Macro for creating [`gio::Action`]s and registering them in a given
+/// [`gio::ActionMap`]. It generates a method `register_actions` with
+/// a following signature:
+///
+/// ```rust
+/// # struct X; impl X {
+/// fn register_actions<AM: glib::object::IsA<gio::ActionMap>>(&self, map: &AM)
+/// # {} }
+/// ```
+///
+/// ## Name of an action
+///
+/// Name of an action is the same as the name of a function. This can be
+/// changed by an annotation.
+///
+/// ```rust
+/// # use gio::prelude::ActionMapExt;
+/// # #[derive(glib::Downgrade)]
+/// # pub struct X;
+/// # #[gio_macros::actions]
+/// # impl X {
+/// // Name of the action is "play".
+/// fn play(&self) {
+/// }
+///
+/// // Name of the action is "pause".
+/// #[action(name = "pause")]
+/// fn pause_handler(&self) {
+/// }
+/// # }
+/// ```
+///
+/// ## Stateful action
+///
+/// An action may have a state. A handler need to be annotated accordingly.
+/// A handler must have a following signature:
+///
+/// ```rust,ignore
+/// #[action(stateful, initial_state = <initial-value>)]
+/// fn action_handler(&self, state: StateType) -> Option<StateType>
+/// ```
+///
+/// `StateType` must implement [`glib::variant::StaticVariantType`]. Returning
+/// `Some(value)` from such handler triggers a change of an action's state.
+///
+/// An `initial_state` annotation may be omited. In this case `StateType` must
+/// implement [`std::default::Default`], which is used to initialize an action.
+///
+/// ## Parameter
+///
+/// An action may have a parameter. In this case a handler need to have one of
+/// the following signatures:
+///
+/// ```rust,ignore
+/// // Stateless action
+/// fn action_handler(&self, parameter: ParameterType)
+///
+/// // Stateful action
+/// #[action(stateful)]
+/// fn action_handler(&self, state: StateType, parameter: ParameterType) -> Option<StateType>
+/// ```
+///
+/// `ParameterType` must also implement [`glib::variant::StaticVariantType`].
+///
+/// ## Change state handler
+///
+/// A handler may be annotated as `change_state`. In this case it will be bound
+/// to a `change-state` signal of an action (in opposite to an `activate`
+/// signal used by default). In this case a signature should look like this:
+///
+/// ```rust,ignore
+/// // Stateful action
+/// #[action(stateful, change_state)]
+/// fn change_state_action_handler(&self, state: StateType) -> Option<StateType>
+/// ```
+///
+/// If you need an action with both `activate` and `change-state` handlers, just
+/// create two methods and annotate them with the same action name.
+///
+/// ## Defaults
+///
+/// A `stateful` annotation may be omited if `initial_state` or `change_state` is
+/// specified.
+///
```

Comment:
Maybe put a note in here about the caveats with async actions? Especially as it relates to returning state changes, it seems like most of the time a stateful action shouldn't be async. You can put a note similar to this one I put in the template callbacks macro: https://github.com/gtk-rs/gtk4-rs/blob/86c1548cc2a6fa05c4285802f0607b374f3bec0f/gtk4-macros/src/lib.rs#L122-L131

Reply:
Why? Such actions work as expected. Returning a value is just a fancy way to invoke `set_state`.

Reply:
Right, but it is possible with async actions to have two or more invocations of the action get spawned. If a later future resolves quicker than an earlier one, that would be like calling `set_state` with a stale value.